### PR TITLE
Misc stuff

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: contrib/epee/include/net/abstract_tcp_server.h [hunk 1]
```diff
@@ -305,7 +305,7 @@ namespace net_utils
 		m_connections.back().powner = this;
 		m_connections.back().m_self_it = --m_connections.end();
 		m_connections.back().m_context.m_remote_address = remote_address;
-		m_connections.back().m_htread = threads_helper::create_thread(ConnectionHandlerProc, &m_connections.back());
+		m_connections.back().m_htread = threads_helper::create_thread(ConnectionHandlerProc, &m_connections.back()); // ugh, seems very risky
```
Comment (stoffu) [inline contrib/epee/include/net/abstract_tcp_server.h]: Why is this risky? Can `m_connections` become empty at this point?

File: src/wallet/wallet_rpc_server.h [hunk 2]
```diff
@@ -129,7 +129,7 @@ namespace tools
       bool on_create_account(const wallet_rpc::COMMAND_RPC_CREATE_ACCOUNT::request& req, wallet_rpc::COMMAND_RPC_CREATE_ACCOUNT::response& res, epee::json_rpc::error& er);
       bool on_label_account(const wallet_rpc::COMMAND_RPC_LABEL_ACCOUNT::request& req, wallet_rpc::COMMAND_RPC_LABEL_ACCOUNT::response& res, epee::json_rpc::error& er);
       bool on_getheight(const wallet_rpc::COMMAND_RPC_GET_HEIGHT::request& req, wallet_rpc::COMMAND_RPC_GET_HEIGHT::response& res, epee::json_rpc::error& er);
-      bool validate_transfer(const std::list<wallet_rpc::transfer_destination>& destinations, const std::string& payment_id, std::vector<cryptonote::tx_destination_entry>& dsts, std::vector<uint8_t>& extra, epee::json_rpc::error& er);
+      bool validate_transfer(const std::list<wallet_rpc::transfer_destination>& destinations, const std::string& payment_id, std::vector<cryptonote::tx_destination_entry>& dsts, std::vector<uint8_t>& extra, bool at_least_one_destination, epee::json_rpc::error& er);
```
Comment (stoffu) [inline src/wallet/wallet_rpc_server.h]: This change seems to deserve a separate commit?

File: contrib/epee/include/net/abstract_tcp_server.h (hunk 1 above)
Comment (julianknutsen) [reply contrib/epee/include/net/abstract_tcp_server.h]: If I'm reading it right, the m_connections object isn't guarded by a lock in this section of the code so it is entirely possible that another thread_context is added to m_connections at any point during lines 304-308 which would cause the back() calls to return different objects.

In addition, the ConnectionHandlerProc, in another thread, could potentially erase() the thread_context (line 154) prior to line 308 running.

Allocating a thread_context() from the heap would allow you to keep the pointer in the list and just use remove() instead of erase(). That way, you can keep the m_connections longer.

Using a condition variable in ConnectionHandlerProc and signaling before invoke_connection() returns would solve the second race.

Not entirely up to speed with c++containers, but that is what I see.

File: contrib/epee/include/net/http_protocol_handler.inl [hunk 3]
```diff
@@ -345,7 +345,12 @@ namespace net_utils
 		{
 			analize_http_method(result, m_query_info.m_http_method, m_query_info.m_http_ver_hi, m_query_info.m_http_ver_hi);
 			m_query_info.m_URI = result[10];
-      parse_uri(m_query_info.m_URI, m_query_info.m_uri_content);
+			if (!parse_uri(m_query_info.m_URI, m_query_info.m_uri_content))
+			{
+				m_state = http_state_error;
```
Comment (julianknutsen) [inline contrib/epee/include/net/http_protocol_handler.inl]: Not sure where the dev team stands on macros w/ return statements (in this case it could also just be another function), but this pattern is used enough in this file it could be cleaned up with something like FAIL_CONNECTION(<reason>) that handles the m_state state and appropriate logging.

File: contrib/epee/src/mlog.cpp [hunk 4]
```diff
@@ -55,6 +55,7 @@ static std::string generate_log_filename(const char *base)
     strcpy(tmp, "unknown");
   else
     strftime(tmp, sizeof(tmp), "%Y-%m-%d-%H-%M-%S", &tm);
+  tmp[sizeof(tmp) - 1] = 0;
```
Comment (julianknutsen) [inline contrib/epee/src/mlog.cpp]: I think strcpy and strftime both null-terminate the buffer. What is the point of setting the last byte to 0 here?

File: external/easylogging++/easylogging++.cc [hunk 5]
```diff
@@ -1015,8 +1015,9 @@ const std::string OS::getBashOutput(const char* command) {
   char hBuff[4096];
   if (fgets(hBuff, sizeof(hBuff), proc) != nullptr) {
     pclose(proc);
-    if (hBuff[strlen(hBuff) - 1] == '\n') {
```
Comment (julianknutsen) [inline external/easylogging++/easylogging++.cc]: Can you help me understand the policies around external/ libraries? If we change this should a PR be sent to the original maintainer? Is it important that we keep these "pure"?

File: src/blockchain_db/blockchain_db.cpp [hunk 6]
```diff
@@ -282,7 +282,7 @@ block BlockchainDB::get_block_from_height(const uint64_t& height) const
   blobdata bd = get_block_blob_from_height(height);
   block b;
   if (!parse_and_validate_block_from_blob(bd, b))
-    throw new DB_ERROR("Failed to parse block from blob retrieved from the db");
+    throw DB_ERROR("Failed to parse block from blob retrieved from the db");
```
[the remark this answers is not in the document]
Comment (julianknutsen) [reply src/blockchain_db/blockchain_db.cpp]: src/cryptonote_core/blockchain.cpp::473 also has the same issue.

File: src/cryptonote_basic/cryptonote_basic.h [hunk 7]
```diff
@@ -259,7 +259,7 @@ namespace cryptonote
             ar.tag("rctsig_prunable");
             ar.begin_object();
             r = rct_signatures.p.serialize_rctsig_prunable(ar, rct_signatures.type, vin.size(), vout.size(),
-                vin[0].type() == typeid(txin_to_key) ? boost::get<txin_to_key>(vin[0]).key_offsets.size() - 1 : 0);
```
Comment (julianknutsen) [inline src/cryptonote_basic/cryptonote_basic.h]: Line 251 already checks for vin.empty(). Can it change in the meantime?

File: src/cryptonote_core/cryptonote_core.cpp [hunk 8]
```diff
@@ -467,7 +467,7 @@ namespace cryptonote
     m_blockchain_storage.set_user_options(blocks_threads,
         blocks_per_sync, sync_mode, fast_sync);
 
-    r = m_blockchain_storage.init(db, m_testnet, test_options);
+    r = m_blockchain_storage.init(db.release(), m_testnet, test_options);
```
Comment (julianknutsen) [inline src/cryptonote_core/cryptonote_core.cpp]: I don't think this is entirely sufficient. Inside Blockchain::init() there is an error path (Line 321) that doesn't free up the memory. My understanding is after Blockchain::init()::327 there is a guarantee that the deinit() function will be called and clean it up. 

Am I following that correctly?

File: src/cryptonote_core/blockchain.cpp [hunk 9]
```diff
@@ -2372,7 +2372,7 @@ bool Blockchain::check_tx_inputs(transaction& tx, uint64_t& max_used_block_heigh
   TIME_MEASURE_FINISH(a);
   if(m_show_time_stats)
```
Comment (julianknutsen) [inline src/cryptonote_core/blockchain.cpp]: These blocks of code are almost identical, but not quite. If they are meant to show the same information, it might make sense to deduplicate it in a maybe_show_time_stats(<tx>) function that has it all in one place. Easier to maintain at least.

File: src/ringct/rctOps.cpp [hunk 10]
```diff
@@ -350,6 +351,7 @@ namespace rct {
     //This takes the outputs and commitments
     //and hashes them into a 32 byte sized key
     key cn_fast_hash(const ctkeyV &PC) {
+        if (PC.empty()) return rct::hash2rct(crypto::cn_fast_hash("", 0));
```
Comment (julianknutsen) [inline src/ringct/rctOps.cpp]: Are there ever expected cases where PC is really empty? Or, should these be asserts? I'm not sure if this is a bugfix or defensive.

File: src/wallet/wallet_rpc_server.h (hunk 2 above)
Comment (moneromooo-monero) [reply src/wallet/wallet_rpc_server.h]: Because it's more involved than just a "if (cond) return;" type line ? It's still a "enforce non empty before dereferencing". Or another reason ?

File: contrib/epee/include/net/http_protocol_handler.inl (hunk 3 above)
Comment (moneromooo-monero) [reply contrib/epee/include/net/http_protocol_handler.inl]: It's a bit subjective, and depends on the case. In general, I try not to change code gratuitously.

File: external/easylogging++/easylogging++.cc (hunk 5 above)
Comment (moneromooo-monero) [reply external/easylogging++/easylogging++.cc]: A PR to the original maintainer is best, yes.

File: src/cryptonote_basic/cryptonote_basic.h (hunk 7 above)
Comment (moneromooo-monero) [reply src/cryptonote_basic/cryptonote_basic.h]: I don't think so.

File: src/cryptonote_core/cryptonote_core.cpp (hunk 8 above)
Comment (moneromooo-monero) [reply src/cryptonote_core/cryptonote_core.cpp]: You seem to be correct, thanks.

File: src/ringct/rctOps.cpp (hunk 10 above)
Comment (moneromooo-monero) [reply src/ringct/rctOps.cpp]: I doubt there are cases where it can be, but I wouldn't bet there won't ever.

File: src/wallet/wallet_rpc_server.cpp [hunk 11]
```diff
@@ -599,6 +606,13 @@ namespace tools
       uint64_t mixin = m_wallet->adjust_mixin(req.mixin);
       std::vector<wallet2::pending_tx> ptx_vector = m_wallet->create_transactions_2(dsts, mixin, req.unlock_time, req.priority, extra, req.account_index, req.subaddr_indices, m_trusted_daemon);
 
+      if (ptx_vector.empty())
+      {
+        er.code = WALLET_RPC_ERROR_CODE_TX_NOT_POSSIBLE;
+        er.message = "No transaction created";
+        return false;
+      }
+
```
Comment (stoffu) [inline src/wallet/wallet_rpc_server.cpp]: Doesn't the same check apply to `transfer_split`?

File: src/wallet/wallet_rpc_server.h (hunk 2 above)
Comment (stoffu) [reply src/wallet/wallet_rpc_server.h]: > It's still a "enforce non empty before dereferencing".

But `dsts` being empty is not dereferenced here nor in wallet2::create_transactions_2 (empty `dsts` causes `error::zero_destination`). As such, isn't this change unnecessary?

Comment (moneromooo-monero) [reply src/wallet/wallet_rpc_server.h]: It is in wallet_rpc_server, in on_sweep_all.

File: src/wallet/wallet_rpc_server.cpp (hunk 11 above)
Comment (moneromooo-monero) [reply src/wallet/wallet_rpc_server.cpp]: The other ones seems fine. This one would have not crashed, but emitted an incorrect message. I think the others are fine, but I might have missed one.

File: src/wallet/wallet_rpc_server.h (hunk 2 above)
Comment (stoffu) [reply src/wallet/wallet_rpc_server.h]: Ah, I see it now. Thanks for clarifying.
